### PR TITLE
Domain adaptation across cancer types

### PR DESCRIPTION
**Pull request by jjc2718**
This PR adds code to test a few simple domain adaptation methods ([CORAL](https://arxiv.org/abs/1612.01939) and [TCA](https://www.ijcai.org/Proceedings/09/Papers/200.pdf)) on mutation prediction across cancer types. The idea for these methods is to apply an unsupervised DA algorithm to align the train data to the test data, then train our models on the aligned training data and evaluate on the test data.

I can't remember if WENDA transforms the data this way (I think it learns a set of feature weights, but I don't remember the details), but I think you should just be able to call your code with the same train/test data that we're using in CORAL and TCA.

Main code changes:

* `02_cancer_type_classification/domain_adaptation.ipynb`: analysis notebook for DA results
* `02_cancer_type_classification/run_cancer_type_classification.py`: add flags for CORAL and TCA parameters (this is a bit of a mess but it works...)
* `pancancer_evaluation/utilities/tcga_utilities.py`: functions to call domain adaptation code in Python [`transfertools`](https://github.com/Vincent-Vercruyssen/transfertools) package

Most other changes are just boilerplate or things copied over from the `mpmp` repo.

### CODE REVIEW

File: 02_cancer_type_classification/domain_adaptation.ipynb [hunk 1]
```diff
@@ -0,0 +1,1567 @@
...
+    "### Simulated data\n",
+    "\n",
+    "How do domain adaptation algorithms available in `transfertools` scale with the number of samples/features?"
+   ]
+  },
+  {
+   "cell_type": "raw",
+   "metadata": {},
+   "source": [
+    "def coral_samples(n_samples, n_features, tol=1e-8, seed=1):\n",
```
Comment (arielah) [inline 02_cancer_type_classification/domain_adaptation.ipynb:42]: Something went weird in the processing of this chunk, and it's pretty hard to read in the notebook form.

File: 02_cancer_type_classification/domain_adaptation.ipynb [hunk 2]
```diff
@@ -0,0 +1,1567 @@
...
...1qpCn3HzyQjNav9gi5189+Mjxqbcy1aMzIy2LJlC8uWLcPKyoqkpCT27NmDrW3eM/s1bdqUffv20b17dwICAvDz88u2f/z48VhbW/Pjjz/KBExCCCFEPsnYXDgVHa14q4mHIn0vOBOlSL/Fyd+/Ha1bt2XHjp2EhgYz53wAbStW5s9rVwHo3r0n06fPUTilEKKomHNd1v/KtWjt2LEjzz33HDNnzsTHx4chQ4bka1AE6NKlC0eOHGHAgAEYjUamTp3Kpk2bSElJoX79+qxevZrmzZvz+uuvA/Daa6/RpUuXovmOhBBCiFJKxmZh6dRqNRMnfsVbbw1iffBN1gffBMDdzZ1Roz5ROJ0QoqTKtWh97bXX2Lx5M2FhYbzwwguPNeObWq1m0qRJ2bb5+vqavr5y5cp/DxFCCCFEHmRsFiVBjx492bRpBwsW/Mzt27dp1Kgxw4ePpGpVL6WjCSFKqFyL1rfffpu3336b48ePs2rVKi5cuMDMmTPp3bv3Q5cUCSGEEML8LHls1qfFkhi0q0DHGnRZE0mptXaF6h+UufRXPKxFi5a0aNFS6RhCiFIi16L1vieffJInn3yShIQENmzYwOjRo1m/fn0xRBNCCCFETixtbPbyKtwMwCEhwQB4exWm6PQodA4hhBCWKc+i9T5nZ2cGDRrEoEGDzJlHCCGEEPlkKWPzwIGDC3X8tGkTARg79ssiSFO0ohNT+X1vAP9cC8Pe2ooezWrwYqs6aErpZFVCiLIhPTqCkM3LCny8LiVrNnqtfcHWk02PjgCX/M/kn++iVQghhBCiLElKy2DI/M3cik4wbbscFsWN8Fg+7+evYDIhhCi4orgqJSQuGgDvShUL1oCL02PlkKJVCCGEECIHG08Ecis6gZqezkzt04SwuBQ+WXOKdcev8nr7hlQp56x0RCGEeGyFvToGiv8KGbm2RQghhBAiB5fDstZWfe2p6jT1KkfPhlVp7VsegCth0UpGE0KIMkWKViGEEEKIHFRwybpX60DgPTJ0Bu7Gp3D+diwAFd0clYwmhBBlilweLIQQQgiRg+db1mb5kUvsvnKXFtM2k5apR2cw0tDLk3pVZHkdIYQoLlK0CiGEEELk4Al3J757sytfbzjKjfBYVCroUM+bsX3boFKplI4nhLBgEVEpLF1/uUDHJqdkAuBgb1XgvquVslvupWgVQgghhMhF0
```
Comment (arielah) [inline 02_cancer_type_classification/domain_adaptation.ipynb:822]: I'm surprised that the values of AUPR(signal) - AUPR(shuffled) would be consistently higher than 0.5, because my intuition tells me that AUPR(shuffled) should be around 0.5. Am I thinking about this wrong?

Comment (arielah) [inline 02_cancer_type_classification/domain_adaptation.ipynb:1483]: What do the dots represent in these plots? They're not outliers, obviously, are they the only 3 observations? If so, how are you getting the box-and-whisker plots you do?

File: 01_stratified_classification/run_stratified_classification.py [hunk 3]
```diff
@@ -120,8 +120,7 @@ def process_args():
                                                 shuffle_labels=shuffle_labels)
                 tcga_data.process_data_for_gene(gene, classification,
                                                 gene_dir,
-                                                use_pancancer=True,
-                                                shuffle_labels=shuffle_labels)
+                                                use_pancancer=True)
```
[the remark this answers is not in the document]
Comment (arielah) [reply 01_stratified_classification/run_stratified_classification.py:123]: Also, why did you get rid of the shuffle_labels parameter? Seems useful for your negative control?

File: pancancer_evaluation/utilities/classify_utilities.py [hunk 4]
```diff
@@ -172,7 +174,13 @@ def run_cv_cancer_type(data_model,
                        num_folds,
                        use_pancancer,
                        use_pancancer_only,
-                       shuffle_labels):
+                       shuffle_labels,
```
Comment (arielah) [inline pancancer_evaluation/utilities/classify_utilities.py:177]: I think this is where shuffle_labels is getting called and throwing the errors in checks?

File: pancancer_evaluation/utilities/classify_utilities.py [hunk 5]
```diff
@@ -734,3 +767,17 @@ def summarize_results_cc(results, train_identifier, test_identifier, signal,
 
     return metrics_out_, roc_df_, pr_df_
 
+
+@contextlib.contextmanager
+def temp_seed(cntxt_seed):
```
Comment (arielah) [inline pancancer_evaluation/utilities/classify_utilities.py:772]: Smart!

File: 01_stratified_classification/run_stratified_classification.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jjc2718) [reply 01_stratified_classification/run_stratified_classification.py:123]: Good catch! I didn't mean to get rid of it entirely - in the `mpmp` repo, I switched to doing the shuffling after I split the data into train/test sets, to make sure the train and test sets have equal proportions of positive/negative samples for the negative control and model trained on true labels. I guess I tried to clean up this repo to be the same, but didn't fix the tests (this is apparently what happens when you edit the code and don't run the tests or create a PR for 5 months...)

I'll fix the tests to work with the new shuffling scheme.

File: 02_cancer_type_classification/domain_adaptation.ipynb (hunk 1 above)
Comment (jjc2718) [reply 02_cancer_type_classification/domain_adaptation.ipynb:42]: Ahh, I had commented out these chunks because they take a while to run. I'll uncomment them and rerun the notebook.

File: 02_cancer_type_classification/domain_adaptation.ipynb (hunk 2 above)
Comment (jjc2718) [reply 02_cancer_type_classification/domain_adaptation.ipynb:1483]: Yeah, each of them is only 4 observations (1 iteration of 4-fold cross-validation). I assume Seaborn estimates the median and IQR by doing some sort of linear interpolation (that's the default in `np.quantile` which I'm guessing Seaborn uses), although I haven't looked into the details.

Comment (jjc2718) [reply 02_cancer_type_classification/domain_adaptation.ipynb:912]: AUROC is usually around 0.5 for a (binary) classifier that guesses randomly, but AUPR can be considerably lower depending on the proportion of positive labels (see, e.g. [this post](https://stats.stackexchange.com/a/266989) for more detail).

Looks like it's around 0.1-0.2 for CDKN2A, for most of the baseline models.

**Review by arielah — APPROVED** [review]
Looks good!
